File: app/src/main/java/ru/vegax/xavier/miniMonsterX/IOData/IOFragment.java
package ru.vegax.xavier.miniMonsterX.IOData;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Objects;

import ru.vegax.xavier.miniMonsterX.IOData.dataLoader.DataLoader;
import ru.vegax.xavier.miniMonsterX.R;

import static ru.vegax.xavier.miniMonsterX.MainActivity.MY_PREFS;


public class IOFragment extends Fragment {
    private static final String TAG = "IOFragment";
    private static final String ARG_URL_JSON = "URL_EXTRA_JSON";
    private static final String ARG_URL_HTML = "URL_EXTRA_HTML";
    private static final String ARG_URL_BASE = "URL_EXTRA_BASE";
    private static final String _urlSufixImpulse = "/?rst=";
    private static final String _urlSufixSet = "/?sw=";

    public static final String PREFF_IMPULSE = "PREFF_IMPULSE";


    private String _urlJSon;
    private String _urlHtml;
    private String _urlBase;

    private ProgressBar _progBarDownload;
    private ArrayList<IOItem> _ioData;
    private IOAdapter _adapter;
    private TextView _txtVdeviceName;
    private DataLoader _dataLoader;
    RecyclerView _recyclerView;

    public IOFragment() {
        // Required empty public constructor
    }

    public static IOFragment newInstance(String urlJSon, String urlHtml, String urlBase) {
        IOFragment fragment = new IOFragment();
        Bundle args = new Bundle();
        args.putString(ARG_URL_JSON, urlJSon);
        args.putString(ARG_URL_HTML, urlHtml);
        args.putString(ARG_URL_BASE, urlBase);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            _urlJSon = getArguments().getString(ARG_URL_JSON);
            _urlHtml = getArguments().getString(ARG_URL_HTML);
            _urlBase = getArguments().getString(ARG_URL_BASE);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View v = inflater.inflate(R.layout.io_data_fragment, container, false);
        _progBarDownload = v.findViewById(R.id.progBarDownloadJS);

        _progBarDownload.setProgress(100);


        _txtVdeviceName = v.findViewById(R.id.txtVdeviceName);

        //Initialize the RecyclerView
        _recyclerView = v.findViewById(R.id.recyclerView);

        //Set the Layout Manager
        _recyclerView.setLayoutManager(new LinearLayoutManager(v.getContext()));

        //Initialize the ArrayLIst that will contain the data
        _ioData = new ArrayList<>();

        //Initialize the adapter and set it ot the RecyclerView

        _adapter = new IOAdapter(v.getContext(), _ioData) {
            @Override
            public boolean onLongClick(View v) {
                impulseOutput(v);
                return true;
            }

            @Override
            public void onClick(View v) {
                setOutput(v);
            }

        };


        _recyclerView.setAdapter(_adapter);
        SharedPreferences sharedPreferences = v.getContext().getSharedPreferences(MY_PREFS, Context.MODE_PRIVATE);

        _dataLoader = new DataLoader(_ioData, sharedPreferences, _urlJSon, _urlHtml, _urlBase) {
            @Override
            public void notifyDataChanged() {
                boolean isButtonChanging = false;
                for (int i = 0; i < DataLoader.PORT_NUMBER; i++) {

                    isButtonChanging |= _ioData.get(i).isChanging(); // if at least one element is being edited

                }
                if (!isButtonChanging) {
                    _adapter.notifyDataSetChanged();
                }
                _progBarDownload.setIndeterminate(false);
                _recyclerView.setVisibility(View.VISIBLE);


            }

            @Override
            public void notifyError(String e) {
                Toast.makeText(v.getContext(), getString(R.string.no_conn), Toast.LENGTH_SHORT).show();
                _progBarDownload.setIndeterminate(false);
                _recyclerView.setVisibility(View.INVISIBLE);
                _txtVdeviceName.setText(getString(R.string.no_conn));
            }

            @Override
            public void setDeviceId(String deviceId) {
                v.post(() -> {
                    _txtVdeviceName.setText(deviceId);//send to a UI thread!
                });

            }

        };
        refreshData();

        return v;
    }


    @Override
    public void onDestroy() {
        _dataLoader.cancelTasks();

        super.onDestroy();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }

    public void setConnData(String urlJSon, String urlHtml, String urlBase) {
        _urlJSon = urlJSon;
        _urlHtml = urlHtml;
        _urlBase = urlBase;
        _dataLoader.setConnData(urlJSon, urlHtml, urlBase);
    }

    //handle click from item

    public void setOutput(View view) {
        int curPos = (int) view.getTag();
        IOItem currentItem = _ioData.get(curPos);
        String suffix;
        if (currentItem.isImpulse()) {
            // set the suffix string for pressing turning the output on or off for a period of time
            suffix = _urlSufixImpulse + (curPos + 1); //
            currentItem.setChanging(false);
            _dataLoader.setOutputs(suffix);
            ((Switch) view).setChecked(currentItem.isOn());
        } else {
            ((Switch) view).setChecked(currentItem.isOn());
            // set the suffix string for pressing turning the output on or off
            suffix = _urlSufixSet + (curPos + 1) + "-" + (currentItem.isOn() ? "0" : "1"); //"/?sw=i-1" for turning on i - port number starting from 1
            currentItem.setChanging(false);
            _dataLoader.setOutputs(suffix);

        }

    }
    //handle long click from item

    public void impulseOutput(View view) {
        int curPos = (int) view.getTag();
        IOItem currentItem = _ioData.get(curPos);

        currentItem.setImpulse(!currentItem.isImpulse());
        SharedPreferences preferences = Objects.requireNonNull(getContext()).getSharedPreferences(MY_PREFS, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = preferences.edit();
        editor.putBoolean(PREFF_IMPULSE + _urlBase + curPos, currentItem.isImpulse());
        editor.apply();
        _adapter.notifyItemChanged(curPos);
        currentItem.setChanging(false);
    }

    public void stopUpdating() {
        _progBarDownload.setIndeterminate(false);
        _dataLoader.stopUptading();
    }

    public void refreshData() {
        if (!_urlHtml.equals("")) {
            _progBarDownload.setIndeterminate(true);

            _dataLoader.refreshData();
        } else {
            _ioData.clear();
            _recyclerView.notify();
        }
    }

    public void clearData() {
        _ioData.clear();
        synchronized (_recyclerView) {
            _recyclerView.notify();
        }
    }

    public void cancelTasks() {
        _dataLoader.stopTasks();
        _progBarDownload.setIndeterminate(false);
    }
}

File: app/src/main/java/ru/vegax/xavier/miniMonsterX/iodata/IOFragment.kt
package ru.vegax.xavier.miniMonsterX.iodata

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import ru.vegax.xavier.miniMonsterX.R
import ru.vegax.xavier.miniMonsterX.activities.IODataViewModel
import ru.vegax.xavier.miniMonsterX.activities.MainActivity.Companion.PREFF_DEV_ID
import ru.vegax.xavier.miniMonsterX.repository.LoadingStatus
import java.util.*


class IOFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {


    private lateinit var mIoData: ArrayList<IOItem>
    private lateinit var mAdapter: IOAdapter

    private lateinit var mRecyclerView: RecyclerView
    private val viewModel by lazy {
        ViewModelProviders.of(activity!!).get(IODataViewModel::class.java)
    }
    private lateinit var swipeContainer: SwipeRefreshLayout

    private var resetDataSet: Boolean = true

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.io_data_fragment, container, false)

        swipeContainer = v.findViewById(R.id.swipe_container)
        swipeContainer.setOnRefreshListener(this)

        mRecyclerView = v.findViewById(R.id.recyclerView)

        mRecyclerView.layoutManager = LinearLayoutManager(v.context)

        mIoData = ArrayList()


        mAdapter = object : IOAdapter(v.context, mIoData) {
            override fun onLongClick(v: View?): Boolean {

                if (v != null) {
                    showNameDialog(v.tag as Int)
                }
                return true
            }

            override fun onClick(v: View) {
                if (v is Switch) {
                    setOutput(v)

                } else if (v is ImageView) {
                    impulseOutput(v)
                }
            }

        }

        mRecyclerView.adapter = mAdapter


        return v
    }

    private fun showNameDialog(curPos: Int) {
        val curContext = context
        if (curContext != null) {
            val d = Dialog(curContext)
            d.setContentView(R.layout.dialog_set_name)
            val btnOk = d.findViewById(R.id.btnSet) as TextView
            val btnCancel = d.findViewById(R.id.btnCancel) as TextView
            val portName = d.findViewById(R.id.edTextPortName) as EditText
            portName.setText(mIoData[curPos].itemName)
            btnOk.setOnClickListener {
                val name = portName.text.toString()
                if (name != "") {
                    val curDevice = viewModel.curDevice
                    if (curDevice != null) {
                        val currentItem = mIoData[curPos]
                        currentItem.itemName = name
                        curDevice.portNames[curPos] = name
                        viewModel.update(curDevice)
                        mAdapter.notifyItemChanged(curPos)
                    }
                    d.dismiss()
                } else {
                    Toast.makeText(curContext, getString(R.string.empty_name), Toast.LENGTH_SHORT).show()
                }
            }
            btnCancel.setOnClickListener {
                d.dismiss()
            }
            d.show()
        }
    }

    override fun onResume() {
        super.onResume()
        observeViewModel()
        getPrefs()

    }

    private fun getPrefs() {
        resetDataSet = true
        val preferences = activity?.getSharedPreferences(MY_PREFS, Context.MODE_PRIVATE)
        val currDeviceID = preferences?.getLong(PREFF_DEV_ID, 0L)
        currDeviceID?.let { viewModel.getDevice(it) }
    }

    private fun rememberPrefs(homeId: Long) {
        val preferences = activity?.getSharedPreferences(MY_PREFS, Context.MODE_PRIVATE)
        val editor = preferences?.edit()
        editor?.putLong(PREFF_DEV_ID, homeId)
        editor?.apply()
    }

    private fun observeViewModel() {
        viewModel.controlData.observe(this, Observer { item ->


            val difference = mIoData.mapIndexed { i, ioItem ->
                ioItem.isOn.xor(item[i].isOn) || ioItem.isOutput.xor(item[i].isOutput)
            }
            mIoData.clear()
            mIoData.addAll(item)
            difference.forEachIndexed { i, different ->
                if (different) mAdapter.notifyItemChanged(i)
            }
            if (resetDataSet) {
                resetDataSet = false
                mAdapter.notifyDataSetChanged()
            }
        })
        viewModel.currentDeviceLiveData.observe(this, Observer { curDevice ->
            if (curDevice != null) {
                resetDataSet = true
                rememberPrefs(curDevice.deviceId)
            }
        })
        viewModel.loadingStatus.observe(this, Observer { loadingStatus ->
            if (loadingStatus != null) {
                when (loadingStatus) {
                    LoadingStatus.LOADING -> swipeContainer.isRefreshing = true
                    LoadingStatus.NOT_LOADING -> swipeContainer.isRefreshing = false
                    LoadingStatus.ERROR -> {
                        updateAdapter()
                    }
                }
            }
        })
    }

    private fun updateAdapter() {
        swipeContainer.isRefreshing = false
        mIoData.clear()
        mAdapter.notifyDataSetChanged()
        Toast.makeText(activity, getString(R.string.cant_update), Toast.LENGTH_SHORT).show()
    }

    override fun onRefresh() {
        refreshData()
    }

    override fun onDestroy() {
        viewModel.stopLoading()
        super.onDestroy()
    }

    fun refreshData() {
        resetDataSet = true
        viewModel.curDevice?.let {
            viewModel.getDataCyclically()
        }
        if (viewModel.curDevice == null) {
            updateAdapter()
        }
    }
    //handle click from item

    fun setOutput(view: View) {
        val url = viewModel.curDevice?.url
        if (url != null) {
            val curPos = view.tag as Int
            val currentItem = mIoData[curPos]
            if (currentItem.isImpulse) {
                currentItem.isChanging = false
                viewModel.setImpulse(curPos + 1)
                (view as Switch).isChecked = currentItem.isOn
            } else {
                (view as Switch).isChecked = currentItem.isOn
                currentItem.isChanging = false
                viewModel.setOutput(curPos + 1, !currentItem.isOn)
            }
        }
    }
    //handle long click from item

    fun impulseOutput(view: View) {
        val curDevice = viewModel.curDevice
        if (curDevice != null) {
            val curPos = view.tag as Int
            val currentItem = mIoData[curPos]
            curDevice.impulseTypes[curPos] = !curDevice.impulseTypes[curPos]
            currentItem.isImpulse = curDevice.impulseTypes[curPos]
            viewModel.update(curDevice)
            mAdapter.notifyItemChanged(curPos)
            currentItem.isChanging = false
        }
    }

    fun stopUpdating() {
        viewModel.stopLoading()
    }


    fun clearData() {
        mIoData.clear()
        synchronized(mRecyclerView) {
            mRecyclerView.adapter?.notifyDataSetChanged()
        }
    }

    fun cancelTasks() {
        viewModel.stopLoading()
    }

    companion object {
        const val TAG = "XavvIOFragment"
        const val MY_PREFS = "my_prefs"

    }
}
